--- src/TerrariaNetCore/Terraria/MessageBuffer.cs
+++ src/tModLoader/Terraria/MessageBuffer.cs
@@ -1,6 +_,7 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics.PackedVector;
 using Terraria.Audio;
@@ -17,12 +_,16 @@
 using Terraria.ID;
 using Terraria.Localization;
 using Terraria.Map;
+using Terraria.ModLoader;
+using Terraria.ModLoader.Exceptions;
+using Terraria.ModLoader.IO;
 using Terraria.Net;
 using Terraria.Net.Sockets;
 using Terraria.Social;
 using Terraria.Social.Steam;
 using Terraria.Testing;
 using Terraria.UI;
+using LegacySoundStyle = Terraria.Audio.SoundStyle;
 
 namespace Terraria;
 
@@ -113,8 +_,14 @@
 		int num = 0;
 		num = start + 1;
 		b = (byte)(messageType = readBuffer[start]);
+
+		if (ModNet.DetailedLogging)
+			ModNet.Debug(whoAmI, $"GetData {MessageID.GetName(b)}({b}), {length+2}");
+
+		/*
 		if (b >= MessageID.Count)
 			return;
+		*/
 
 		Main.ActiveNetDiagnosticsUI.CountReadMessage(b, length);
 		if (Main.netMode == 1 && Netplay.Connection.StatusMax > 0)
@@ -135,7 +_,8 @@
 		}
 
 		if (Main.netMode == 2) {
+			//TML: Added 'b < 250' check.
-			if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b != 147 && b != 161)
+			if (Netplay.Clients[whoAmI].State < 10 && b > 12 && b != 93 && b != 16 && b != 42 && b != 50 && b != 38 && b != 68 && b != 147 && b != 161 && b < 250)
 				NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
 
 			if (Netplay.Clients[whoAmI].State == 0 && b != 1)
@@ -146,6 +_,10 @@
 			ResetReader();
 
 		reader.BaseStream.Position = num;
+
+		if (ModNet.HijackGetData(ref b, ref reader, whoAmI))
+			return;
+
 		switch (b) {
 			case 1: {
 				if (Main.netMode != 2)
@@ -163,9 +_,20 @@
 					if (Netplay.Clients[whoAmI].State != 0)
 						break;
 
+					string clientVersion = reader.ReadString();
+
+					/*
 					if (reader.ReadString() == "Terraria" + 325) {
+					*/
+					if (ModNet.IsClientCompatible(clientVersion, out ModNet.isModdedClient[whoAmI], out var kickMsg)) {
 						if (string.IsNullOrEmpty(Netplay.ServerPassword)) {
 							Netplay.Clients[whoAmI].State = 1;
+
+							if (ModNet.isModdedClient[whoAmI]) {
+								ModNet.SyncMods(whoAmI);
+								return;
+							}
+
 							NetMessage.TrySendData(3, whoAmI);
 						}
 						else {
@@ -174,7 +_,10 @@
 						}
 					}
 					else {
+						/*
 						NetMessage.TrySendData(2, whoAmI, -1, Lang.mp[4].ToNetworkText());
+						*/
+						NetMessage.TrySendData(2, whoAmI, -1, NetworkText.FromLiteral(Lang.mp[4].Value + $"\n({kickMsg})"));
 					}
 				}
 
@@ -184,6 +_,9 @@
 				if (Main.netMode == 1) {
 					Netplay.Disconnect = true;
 					Main.statusText = NetworkText.Deserialize(reader).ToString();
+
+					// Added by TML.
+					Main.menuMode = MenuID.MultiplayerJoining;
 				}
 				break;
 			case 3:
@@ -230,6 +_,9 @@
 					if (!string.IsNullOrWhiteSpace(Netplay.HostToken))
 						NetMessage.TrySendData(161, -1, -1, NetworkText.FromLiteral(Netplay.HostToken));
 
+
+					PlayerLoader.SyncPlayer(player19, -1, -1, true);
+
 					NetMessage.TrySendData(6);
 					if (Netplay.Connection.State == 2)
 						Netplay.Connection.State = 3;
@@ -255,11 +_,11 @@
 
 				player15.voicePitchOffset = Utils.Clamp(player15.voicePitchOffset, -1f, 1f);
 				player15.hair = reader.ReadByte();
-				if (player15.hair >= 228)
+				if (player15.hair >= HairLoader.Count)
 					player15.hair = 0;
 
 				player15.name = reader.ReadString().Trim().Trim();
-				player15.hairDye = reader.ReadByte();
+				player15.hairDye = reader.Read7BitEncodedInt();
 				ReadAccessoryVisibility(reader, player15.hideVisibleAccessory);
 				player15.hideMisc = reader.ReadByte();
 				player15.hairColor = reader.ReadRGB();
@@ -335,9 +_,12 @@
 			case 5: {
 				int num186 = reader.ReadByte();
 				int num187 = reader.ReadInt16();
+				/*
 				int stack6 = reader.ReadInt16();
 				int prefixWeWant3 = reader.ReadByte();
 				int type16 = reader.ReadInt16();
+				*/
+				Item item4 = ItemIO.Receive(reader, readStack: true);
 				BitsByte bitsByte35 = reader.ReadByte();
 				bool favorited = bitsByte35[0];
 				bool flag17 = bitsByte35[1];
@@ -351,10 +_,12 @@
 				lock (player16) {
 					PlayerItemSlotID.SlotReference slot = new PlayerItemSlotID.SlotReference(player16, num187);
 					PlayerItemSlotID.SlotReference slotReference = new PlayerItemSlotID.SlotReference(Main.clientPlayer, num187);
+					/*
 					Item item4 = new Item();
 					item4.SetDefaults(type16);
 					item4.stack = stack6;
 					item4.Prefix(prefixWeWant3);
+					*/
 					item4.favorited = favorited;
 					slot.Item = item4;
 					if (num186 == Main.myPlayer && !Main.ServerSideCharacter)
@@ -566,6 +_,10 @@
 					Main.LobbyId = reader.ReadUInt64();
 					Sandstorm.IntendedSeverity = reader.ReadSingle();
 					ExtraSpawnPointManager.Read(reader, networking: true);
+
+					if (Netplay.Connection.State > 4)
+						WorldIO.ReceiveModData(reader);
+
 					if (Netplay.Connection.State == 3) {
 						Main.windSpeedCurrent = Main.windSpeedTarget;
 						Netplay.Connection.State = 4;
@@ -800,6 +_,10 @@
 					NetMessage.TrySendData(12, -1, whoAmI, null, whoAmI, (int)(byte)playerSpawnContext);
 					NetMessage.TrySendData(129, whoAmI);
 					NetMessage.greetPlayer(whoAmI);
+
+					if (Main.ActiveWorldFileData.ModSaveErrors.Any()) {
+						ChatHelper.SendChatMessageToClient(Utils.CreateSaveErrorMessage("tModLoader.WorldCustomDataSaveFail", Main.ActiveWorldFileData.ModSaveErrors), Color.OrangeRed, whoAmI);
+					}
 					if (Main.player[num192].unlockedBiomeTorches) {
 						NPC nPC6 = new NPC();
 						nPC6.SetDefaults(664);
@@ -943,8 +_,11 @@
 					Player player9 = Main.player[num97];
 					player9.statLife = reader.ReadInt16();
 					player9.statLifeMax = reader.ReadInt16();
+					// Allow health stat to exceed vanilla bounds (#HealthManaAPI)
+					/*
 					if (player9.statLifeMax < 20)
 						player9.statLifeMax = 20;
+					*/
 
 					player9.dead = player9.statLife <= 0;
 					if (Main.netMode == 2)
@@ -1149,7 +_,7 @@
 				BitsByte bitsByte22 = (byte)0;
 				BitsByte bitsByte23 = (byte)0;
 				BitsByte bitsByte24 = (byte)0;
-				Tile tile4 = null;
+				Tile tile4 = default;
 				for (int num90 = num86; num90 < num86 + num88; num90++) {
 					for (int num91 = num87; num91 < num87 + num89; num91++) {
 						if (Main.tile[num90, num91] == null)
@@ -1229,8 +_,12 @@
 				int num39 = reader.ReadInt16();
 				Vector2 vector = reader.ReadVector2();
 				Vector2 velocity = reader.ReadVector2();
+/*
 				int stack = reader.ReadInt16();
 				int prefix = reader.ReadByte();
+*/
+				int stack = reader.Read7BitEncodedInt();
+				int prefix = reader.Read7BitEncodedInt();
 				BitsByte bitsByte3 = reader.ReadByte();
 				bool num40 = bitsByte3[2];
 				bool flag2 = bitsByte3[3];
@@ -1256,6 +_,7 @@
 
 					worldItem.Prefix(prefix);
 					worldItem.stack = stack;
+					ItemIO.ReceiveModData(worldItem.inner, reader);
 					worldItem.position = vector;
 					worldItem.velocity = velocity;
 					worldItem.shimmered = shimmered;
@@ -1268,6 +_,7 @@
 					}
 				}
 				else {
+					if (num39 == 400) Main.timeItemSlotCannotBeReusedFor[num39] = 0; // Fix #4372
 					if (Main.timeItemSlotCannotBeReusedFor[num39] > 0)
 						break;
 
@@ -1287,6 +_,7 @@
 
 					worldItem.Prefix(prefix);
 					worldItem.stack = stack;
+					ItemIO.ReceiveModData(worldItem.inner, reader);
 					worldItem.position = vector;
 					worldItem.velocity = velocity;
 					worldItem.shimmered = shimmered;
@@ -1437,9 +_,13 @@
 				if (num177 > -1)
 					nPC4.TransformVisuals(num177, nPC4.type);
 
-				if (nPC4.type >= 0 && nPC4.type < NPCID.Count && Main.npcCatchable[nPC4.type])
+				if (nPC4.type >= 0 && Main.npcCatchable[nPC4.type])
 					nPC4.releaseOwner = reader.ReadByte();
 
+				// Extra AI is read into a temporary buffer for parity with ProjectileLoader code, and to avoid exceptions causing underreads.
+				if (bitsByte34[5])
+					NPCLoader.ReceiveExtraAI(nPC4, NPCLoader.ReadExtraAI(reader));
+
 				if (flag15)
 					nPC4.OnSpawn(new EntitySource_Sync());
 
@@ -1463,6 +_,8 @@
 				float knockBack2 = (bitsByte28[5] ? reader.ReadSingle() : 0f);
 				int originalDamage = (bitsByte28[6] ? reader.ReadInt16() : 0);
 				array[2] = (bitsByte29[0] ? reader.ReadSingle() : 0f);
+				byte[] extraAI = bitsByte29[1] ? ProjectileLoader.ReadExtraAI(reader) : null;
+
 				if (Main.netMode == 2 && (Main.projHostile[num154] || key2.Spawner != whoAmI))
 					break;
 
@@ -1490,6 +_,10 @@
 					proj2.ai[num155] = array[num155];
 				}
 
+				if (extraAI != null) {
+					ProjectileLoader.ReceiveExtraAI(proj2, extraAI);
+				}
+
 				if (flag13)
 					proj2.FinalizeProjectile();
 
@@ -1504,10 +_,29 @@
 
 				int num207 = reader.ReadByte();
 				int num208 = reader.ReadByte();
+				/*
 				int num209 = reader.ReadInt16();
 				float num210 = reader.ReadSingle();
 				int num211 = reader.ReadByte() - 1;
 				byte b18 = reader.ReadByte();
+				*/
+
+				var num209 = reader.Read7BitEncodedInt();
+				NPC.HitInfo hit = default;
+				if (num209 >= 0) {
+					hit = new NPC.HitInfo {
+						Damage = num209,
+						SourceDamage = reader.Read7BitEncodedInt(),
+						DamageType = DamageClassLoader.DamageClasses[reader.Read7BitEncodedInt()],
+						HitDirection = reader.ReadSByte(),
+						Knockback = reader.ReadSingle()
+					};
+					BitsByte flags = reader.ReadByte();
+					hit.Crit = flags[0];
+					hit.InstantKill = flags[1];
+					hit.HideCombatText = flags[2];
+				}
+
 				NPC nPC7 = Main.npc[num207];
 				if (Main.netMode == 2) {
 					if (nPC7.generation != num208)
@@ -1523,7 +_,10 @@
 				}
 
 				if (num209 >= 0) {
+					/*
 					nPC7.StrikeNPC(num209, num210, num211, b18 == 1, fromNet: true, (Main.netMode == 2) ? whoAmI : 255);
+					*/
+					nPC7.StrikeNPC(hit, fromNet: true, owner: (Main.netMode == 2) ? whoAmI : 255);
 				}
 				else {
 					nPC7.life = 0;
@@ -1532,7 +_,11 @@
 				}
 
 				if (Main.netMode == 2) {
+					/*
 					NetMessage.TrySendData(28, -1, whoAmI, null, num207, num209, num210, num211, b18);
+					*/
+					NetMessage.SendStrikeNPC(nPC7, hit, ignoreClient: whoAmI);
+
 					if (nPC7.life <= 0)
 						NetMessage.TrySendData(23, -1, -1, null, num207);
 
@@ -1610,16 +_,26 @@
 			case 32: {
 				int num149 = reader.ReadInt16();
 				int num150 = reader.ReadByte();
+				/*
 				int stack5 = reader.ReadInt16();
 				int prefixWeWant2 = reader.ReadByte();
 				int type14 = reader.ReadInt16();
+				*/
+				int stack5 = -1;
+				int prefixWeWant2 = -1;
+				int type14 = -1;
 				if (num149 >= 0 && num149 < 8000 && Main.chest[num149] != null) {
 					if (Main.chest[num149].item[num150] == null)
 						Main.chest[num149].item[num150] = new Item();
 
+					ItemIO.Receive(Main.chest[num149].item[num150], reader, readStack: true);
+					goto SkipVanillaItemSetup;
+
 					Main.chest[num149].item[num150].SetDefaults(type14);
 					Main.chest[num149].item[num150].Prefix(prefixWeWant2);
 					Main.chest[num149].item[num150].stack = stack5;
+					SkipVanillaItemSetup:
+
 					if (Main.netMode == 2)
 						NetMessage.TrySendData(32, -1, whoAmI, null, num149, num150);
 				}
@@ -1633,7 +_,7 @@
 				int num24 = reader.ReadByte();
 				string name = string.Empty;
 				if (num24 != 0) {
-					if (num24 <= 20)
+					if (num24 <= Chest.MaxNameLength)
 						name = reader.ReadString();
 					else if (num24 != 255)
 						num24 = 0;
@@ -1687,25 +_,49 @@
 				if (Main.netMode == 2)
 					num32 = 0;
 
+
+				ushort modType = 0;
+				if (b4 >= 100)
+					modType = reader.ReadUInt16();
+
 				if (Main.netMode == 2) {
 					using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
 						switch (b4) {
+							/*
 							case 0: {
+							*/
+							case byte _ when b4 % 100 == 0: {
+								if (modType == 0)
+									modType = TileID.Containers;
+
+								//TML: '21' -> 'modType'
-								int num35 = WorldGen.PlaceChest(num29, num30, 21, notNearOtherChests: false, num31);
+								int num35 = WorldGen.PlaceChest(num29, num30, modType, notNearOtherChests: false, num31);
+
 								if (num35 == -1) {
+									//TML: Added 'modType' at the end.
-									NetMessage.TrySendData(34, whoAmI, -1, null, b4, num29, num30, num31, num35);
+									NetMessage.TrySendData(34, whoAmI, -1, null, b4, num29, num30, num31, num35, modType);
+									/*
 									int itemDrop_Chests2 = WorldGen.GetItemDrop_Chests(num31, secondType: false);
+									*/
+									int itemDrop_Chests2 = b4 < 100 ? WorldGen.GetItemDrop_Chests(num31, secondType: false) : TileLoader.GetItemDropFromTypeAndStyle(modType, num31);
 									if (itemDrop_Chests2 > 0)
 										Item.NewItem(new EntitySource_TileBreak(num29, num30), num29 * 16, num30 * 16, 32, 32, itemDrop_Chests2, 1, noBroadcast: true);
 								}
 								else {
+									//TML: Added 'modType' at the end.
-									NetMessage.TrySendData(34, -1, -1, null, b4, num29, num30, num31, num35);
+									NetMessage.TrySendData(34, -1, -1, null, b4, num29, num30, num31, num35, modType);
 								}
 
 								break;
 							}
+							/*
 							case 1:
+							*/
+							case byte _ when b4 % 100 == 1:
+								/*
 								if (Main.tile[num29, num30].type == 21) {
+								*/
+								if (Main.tile[num29, num30].type == 21 || b4 == 101 && TileID.Sets.BasicChest[Main.tile[num29, num30].type]) {
 									Tile tile = Main.tile[num29, num30];
 									if (tile.frameX % 36 != 0)
 										num29--;
@@ -1723,20 +_,42 @@
 								goto default;
 							default:
 								switch (b4) {
+									/*
 									case 2: {
+									*/
+									case byte _ when b4 % 100 == 2: {
+										if (modType == 0)
+											modType = TileID.Dressers;
+
+										//TML: '88' -> 'modType'
-										int num33 = WorldGen.PlaceChest(num29, num30, 88, notNearOtherChests: false, num31);
+										int num33 = WorldGen.PlaceChest(num29, num30, modType, notNearOtherChests: false, num31);
+
 										if (num33 == -1) {
+											//TML: Added 'modType' at the end.
-											NetMessage.TrySendData(34, whoAmI, -1, null, b4, num29, num30, num31, num33);
+											NetMessage.TrySendData(34, whoAmI, -1, null, b4, num29, num30, num31, num33, modType);
+
+											/*
 											Item.NewItem(new EntitySource_TileBreak(num29, num30), num29 * 16, num30 * 16, 32, 32, WorldGen.GetItemDrop_Dressers(num31), 1, noBroadcast: true);
+											*/
+											int itemSpawn = b4 < 100 ? WorldGen.GetItemDrop_Dressers(num31) : TileLoader.GetItemDropFromTypeAndStyle(modType, num31);
+											if (itemSpawn > 0)
+												Item.NewItem(new EntitySource_TileBreak(num29, num30), num29 * 16, num30 * 16, 32, 32, itemSpawn, 1, noBroadcast: true);
 										}
 										else {
+											//TML: Added 'modType' at the end.
-											NetMessage.TrySendData(34, -1, -1, null, b4, num29, num30, num31, num33);
+											NetMessage.TrySendData(34, -1, -1, null, b4, num29, num30, num31, num33, modType);
 										}
 
 										break;
 									}
+									/*
 									case 3:
+									*/
+									case byte _ when b4 % 100 == 3:
+										/*
 										if (Main.tile[num29, num30].type == 88) {
+										*/
+										if (Main.tile[num29, num30].type == 88 || (b4 == 103 && TileID.Sets.BasicDresser[Main.tile[num29, num30].type])) {
 											Tile tile2 = Main.tile[num29, num30];
 											num29 -= tile2.frameX % 54 / 18;
 											if (tile2.frameY % 36 != 0)
@@ -1792,21 +_,37 @@
 				}
 
 				switch (b4) {
+					/*
 					case 0:
+					*/
+					case byte _ when b4 % 100 == 0:
 						if (num32 == -1) {
 							WorldGen.KillTile(num29, num30);
 							break;
 						}
 						SoundEngine.PlaySound(0, num29 * 16, num30 * 16);
+
+						if (modType == 0)
+							modType = TileID.Containers;
+
+						//TML: '21' -> 'modType'.
-						WorldGen.PlaceChestDirect(num29, num30, 21, num31, num32);
+						WorldGen.PlaceChestDirect(num29, num30, modType, num31, num32);
 						break;
+					/*
 					case 2:
+					*/
+					case byte _ when b4 % 100 == 2:
 						if (num32 == -1) {
 							WorldGen.KillTile(num29, num30);
 							break;
 						}
 						SoundEngine.PlaySound(0, num29 * 16, num30 * 16);
+
+						if (modType == 0)
+							modType = TileID.Dressers;
+
+						//TML: '88' -> 'modType'.
-						WorldGen.PlaceDresserDirect(num29, num30, 88, num31, num32);
+						WorldGen.PlaceDresserDirect(num29, num30, modType, num31, num32);
 						break;
 					case 4:
 						if (num32 == -1) {
@@ -1851,6 +_,12 @@
 				player14.zone4 = reader.ReadByte();
 				player14.zone5 = reader.ReadByte();
 				player14.townNPCs = reader.ReadByte();
+
+				BiomeLoader.ReceiveCustomBiomes(player14, reader);
+
+				//TML: Recalculate the 'not in any biome' value.
+				player14.ZonePurity = player14.InZonePurity();
+
 				if (Main.netMode == 2) {
 					if (!flag10 && player14.zone5[0])
 						NPC.Spawner.SpawnFaelings(player14);
@@ -1876,6 +_,12 @@
 				if (Main.netMode == 2) {
 					if (reader.ReadString() == Netplay.ServerPassword) {
 						Netplay.Clients[whoAmI].State = 1;
+
+						if (ModNet.isModdedClient[whoAmI]) {
+							ModNet.SyncMods(whoAmI);
+							return;
+						}
+
 						NetMessage.TrySendData(3, whoAmI);
 					}
 					else {
@@ -1922,7 +_,12 @@
 				float itemRotation = reader.ReadSingle();
 				int itemAnimation = reader.ReadInt16();
 				player2.itemRotation = itemRotation;
+				// TML changes how item synchronization works, making it based on remote players' already-synchronized input.
+				// Setting itemAnimation in this packet would interfere with that. #ItemTimeOnAllClients
+				/*
 				player2.itemAnimation = itemAnimation;
+				*/
+				
 				player2.channel = player2.inventory[player2.selectedItem].channel;
 				if (Main.netMode == 2)
 					NetMessage.TrySendData(41, -1, whoAmI, null, num28);
@@ -2047,7 +_,11 @@
 				if (Main.tile[num10, num11] == null)
 					Main.tile[num10, num11] = new Tile();
 
+				// Can't lock on struct. Lock does nothing anyway as this is the only place it's acquired
+				/*
 				lock (Main.tile[num10, num11]) {
+				*/
+				{
 					Main.tile[num10, num11].liquid = b2;
 					Main.tile[num10, num11].liquidType(liquidType);
 					if (Main.netMode == 2) {
@@ -2313,7 +_,7 @@
 				if (Main.netMode != 2)
 					break;
 
-				if (num205 >= 0 && num205 < NPCID.Count && NPCID.Sets.MPAllowedEnemies[num205]) {
+				if (num205 >= 0 && NPCID.Sets.MPAllowedEnemies[num205]) {
 					if (!NPC.AnyNPCs(num205))
 						NPC.SpawnOnPlayer(num204, num205);
 				}
@@ -2751,7 +_,8 @@
 						TileEntity.Remove(result3);
 				}
 				else {
-					TileEntity tileEntity = TileEntity.Read(reader, 325, networkSend: true);
+					//TML: Made 'lightSend' true.
+					TileEntity tileEntity = TileEntity.Read(reader, Main.curRelease, networkSend: true, lightSend: true);
 					tileEntity.ID = num115;
 					TileEntity.Add(tileEntity);
 				}
@@ -2827,11 +_,22 @@
 
 				int x15 = reader.ReadInt16();
 				int y15 = reader.ReadInt16();
+
+				Item item;
+				item = ItemIO.Receive(reader);
+				item.stack = reader.Read7BitEncodedInt();
+				/*
 				int type20 = reader.ReadInt16();
 				int prefix5 = reader.ReadByte();
 				int stack8 = reader.ReadInt16();
+				item = new Item(netid3, stack6, prefix3);
+				*/
+
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEItemFrame.TryPlacing(x15, y15, type20, prefix5, stack8);
+					*/
+					TEItemFrame.TryPlacing(x15, y15, item, 1);
 					break;
 				}
 			}
@@ -2841,6 +_,13 @@
 
 				int num213 = reader.ReadInt32();
 				int num214 = reader.ReadByte();
+
+				// If type corresponds to projectile (magic number 2), read another byte for owner #WorldUIAnchorProjectileSyncFix
+				if (num214 == 2) {
+					int owner = reader.ReadByte();
+					num214 |= owner << 8; // Reassign num214 for use in DeserializeNetAnchor - this is now "packedOwnerType"
+				}
+
 				if (num214 == 255) {
 					if (EmoteBubble.byID.ContainsKey(num213))
 						EmoteBubble.byID.Remove(num213);
@@ -2873,6 +_,7 @@
 					EmoteBubble.byID[num213].ID = num213;
 					EmoteBubble.byID[num213].metadata = metadata;
 					EmoteBubble.OnBubbleChange(num213);
+					EmoteBubbleLoader.OnSpawn(EmoteBubble.byID[num213]);
 					break;
 				}
 			}
@@ -3244,7 +_,16 @@
 				break;
 			case 117: {
 				int num25 = reader.ReadByte();
+				bool directHurt = false;
+				if (num25 == 255) {
+					num25 = reader.ReadByte();
+					directHurt = true;
+				}
+
 				if (Main.netMode != 2 || whoAmI == num25 || (Main.player[num25].hostile && Main.player[whoAmI].hostile)) {
+					if (directHurt)
+						goto ReadDirectHurt;
+
 					PlayerDeathReason playerDeathReason2 = PlayerDeathReason.FromReader(reader);
 					int damage = reader.ReadInt16();
 					int num26 = reader.ReadByte() - 1;
@@ -3255,6 +_,26 @@
 					Main.player[num25].Hurt(playerDeathReason2, damage, num26, pvp2, quiet: true, flag, num27);
 					if (Main.netMode == 2)
 						NetMessage.SendPlayerHurt(num25, playerDeathReason2, damage, num26, flag, pvp2, num27, -1, whoAmI);
+
+					break;
+
+					ReadDirectHurt:
+					BitsByte pack = reader.ReadByte();
+					var args = new Player.HurtInfo {
+						DamageSource = PlayerDeathReason.FromReader(reader),
+						PvP = pack[0],
+						CooldownCounter = reader.ReadSByte(),
+						Dodgeable = pack[1],
+						SourceDamage = reader.Read7BitEncodedInt(),
+						Damage = reader.Read7BitEncodedInt(),
+						HitDirection = reader.ReadSByte(),
+						Knockback = reader.ReadSingle(),
+						DustDisabled = pack[2],
+						SoundDisabled = pack[3]
+					};
+					Main.player[num25].Hurt(args, quiet: true);
+					if (Main.netMode == 2)
+						NetMessage.SendPlayerHurt(num25, args, ignoreClient: whoAmI);
 				}
 
 				break;
@@ -3280,7 +_,7 @@
 					num258 = whoAmI;
 
 				int num259 = reader.ReadByte();
-				if (num259 >= 0 && num259 < EmoteID.Count && Main.netMode == 2) {
+				if (num259 >= 0 && num259 < EmoteBubbleLoader.EmoteBubbleCount && Main.netMode == 2) {
 					EmoteBubble.NewBubble(num259, new WorldUIAnchor(Main.player[num258]), 360);
 					EmoteBubble.CheckForNPCsToReactToEmoteBubble(num259, Main.player[num258]);
 				}
@@ -3341,11 +_,21 @@
 
 				int x13 = reader.ReadInt16();
 				int y13 = reader.ReadInt16();
+
+				Item item;
+				item = ItemIO.Receive(reader, readStack: true);
+				/*
 				int type18 = reader.ReadInt16();
 				int prefix4 = reader.ReadByte();
 				int stack7 = reader.ReadInt16();
+				item = new Item(netid2, stack4, prefix2);
+				*/
+
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEWeaponsRack.TryPlacing(x13, y13, type18, prefix4, stack7);
+					*/
+					TEWeaponsRack.TryPlacing(x13, y13, item, item.stack);
 					break;
 				}
 			}
@@ -3474,10 +_,23 @@
 					ushort key = reader.ReadUInt16();
 					LegacySoundStyle legacySoundStyle = SoundID.SoundByIndex[key];
 					BitsByte bitsByte27 = reader.ReadByte();
+					/*
 					int num116 = -1;
 					float num117 = 1f;
 					float num118 = 0f;
 					SoundEngine.PlaySound(Style: (!bitsByte27[0]) ? legacySoundStyle.Style : reader.ReadInt32(), volumeScale: (!bitsByte27[1]) ? legacySoundStyle.Volume : MathHelper.Clamp(reader.ReadSingle(), 0f, 1f), pitchOffset: (!bitsByte27[2]) ? legacySoundStyle.GetRandomPitch() : MathHelper.Clamp(reader.ReadSingle(), -1f, 1f), type: legacySoundStyle.SoundId, x: point2.X, y: point2.Y);
+					*/
+
+					if (bitsByte27[0])
+						legacySoundStyle.Variants = new int[] { reader.ReadInt32() };
+
+					if (bitsByte27[1])
+						legacySoundStyle.Volume = MathHelper.Clamp(reader.ReadSingle(), 0f, 1f);
+
+					if (bitsByte27[2])
+						legacySoundStyle.Pitch = MathHelper.Clamp(reader.ReadSingle(), 0f, 1f);
+
+					SoundEngine.PlaySound(legacySoundStyle, point2.ToVector2());
 				}
 				break;
 			case 133: {
@@ -3486,17 +_,27 @@
 
 				int x8 = reader.ReadInt16();
 				int y8 = reader.ReadInt16();
+
+				Item item;
+				item = ItemIO.Receive(reader, readStack: true);
+				/*
 				int type10 = reader.ReadInt16();
 				int prefix3 = reader.ReadByte();
 				int stack4 = reader.ReadInt16();
+				item = new Item(netid, stack, prefix);
+				*/
+
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEFoodPlatter.TryPlacing(x8, y8, type10, prefix3, stack4);
+					*/
+					TEFoodPlatter.TryPlacing(x8, y8, item, item.stack);
 					break;
 				}
 			}
 			case 134: {
 				int num103 = reader.ReadByte();
-				int ladyBugLuckTimeLeft = reader.ReadInt32();
+				double ladyBugLuckTimeLeft = reader.ReadDouble();
 				float torchLuck = reader.ReadSingle();
 				byte luckPotion = reader.ReadByte();
 				bool hasGardenGnomeNearby = reader.ReadBoolean();
@@ -3641,11 +_,18 @@
 
 				int x3 = reader.ReadInt16();
 				int y3 = reader.ReadInt16();
+				Item item = ItemIO.Receive(reader);
+				int stack = reader.Read7BitEncodedInt();
+				/*
 				int type4 = reader.ReadInt16();
 				int prefix2 = reader.ReadByte();
 				int stack2 = reader.ReadInt16();
+				*/
 				using (Item.DefaultAssignNewItemsToPlayer(whoAmI)) {
+					/*
 					TEDeadCellsDisplayJar.TryPlacing(x3, y3, type4, prefix2, stack2);
+					*/
+					TEDeadCellsDisplayJar.TryPlacing(x3, y3, item, stack);
 					break;
 				}
 			}
@@ -3683,7 +_,10 @@
 					Player player5 = Main.player[num47];
 					Item item2 = player5.inventory[player5.selectedItem];
 					if (item2.UseSound != null)
+						/*
 						SoundEngine.PlaySound(item2.UseSound, player5.Center, item2.useSoundPitch);
+						*/
+						SoundEngine.PlaySound(item2.UseSound, player5.Center);
 				}
 
 				break;
@@ -3744,6 +_,35 @@
 				Main.player[whoAmI].host = !string.IsNullOrWhiteSpace(Netplay.HostToken) && Netplay.HostToken == text;
 				break;
 			}
+			case MessageID.InGameChangeConfig:
+				ModLoader.Config.ConfigManager.HandleInGameChangeConfigPacket(reader, whoAmI);
+				return;
+			case MessageID.ModPacket:
+				ModNet.HandleModPacket(reader, whoAmI, length);
+				return;
+			case MessageID.SyncMods:
+				if (Main.netMode == 1) {
+					ModNet.SyncClientMods(reader);
+				}
+				else {
+					ModNet.SendNetIDs(whoAmI);
+					NetMessage.SendData(3, whoAmI);
+				}
+				return;
+			case MessageID.ModFile:
+				if (Main.netMode == 1)
+					ModNet.ReceiveMod(reader);
+				else
+					ModNet.SendMod(reader.ReadString(), whoAmI);
+				return;
+			case MessageID.KeepAliveDuringModReload:
+				string keepAlive = "Keep Alive During Mod Reload";
+				ModNet.Log(whoAmI, keepAlive);
+				// Logging.ServerConsoleLine doesn't log client info and ModNet.Log doesn't output to console, so manually replicate here.
+				var client = Netplay.Clients[whoAmI];
+				Console.WriteLine($"[{whoAmI}][{client.Socket?.GetRemoteAddress()?.GetFriendlyName()} ({client.Name})]: {keepAlive}");
+				// Does nothing. "Netplay.Clients[whoAmI].TimeOutTimer = 0;" at top of method should keep the connection alive
+				return;
 			default:
 				if (Main.netMode == 2 && Netplay.Clients[whoAmI].State == 0)
 					NetMessage.BootPlayer(whoAmI, Lang.mp[2].ToNetworkText());
