--- src/TerrariaNetCore/Terraria/Testing/ChatCommands/ToolkitDebugCommands.cs
+++ src/tModLoader/Terraria/Testing/ChatCommands/ToolkitDebugCommands.cs
@@ -4,6 +_,7 @@
 using System.IO;
 using Microsoft.Xna.Framework;
 using ReLogic.OS;
+using ReLogic.OS;
 using Terraria.Chat;
 using Terraria.GameContent;
 using Terraria.GameContent.UI.States;
@@ -127,10 +_,12 @@
 			return false;
 		}
 
-		Main.UnpinFromCore0 = !Main.UnpinFromCore0;
+		Main.UnpinFromCore0 ^= true;
 		Main.SaveSettings();
+
-		message.Reply("Core 0 pinning " + ((!Main.UnpinFromCore0) ? "enabled" : "disabled"));
+		message.Reply("Core 0 pinning " + (!Main.UnpinFromCore0 ? "enabled" : "disabled"));
 		WindowsPerformanceDiagnostics.SetCore0Affinity(!Main.UnpinFromCore0);
+
 		return true;
 	}
 
